src/System.Management.Automation/engine/SessionStateLocationAPIs.cs

```diff
@@ -231,6 +231,15 @@ internal PathInfo SetLocation(string path, CmdletProviderContext context)
             ProviderInfo provider = null;
             string providerId = null;
 
+
+            const string locationStackHistoryId = "SetLocationHistoryStack";
+            // Replace path with last working directory when '-' was passed.
+            if (originalPath.Equals("-"))            
+            {
+                return PopLocation(locationStackHistoryId);
+            }
+            PushCurrentLocation(locationStackHistoryId);
```

Comment:
Nice - so simple. I do have a concern about an unbounded stack though. This stack will not be popped often, so I think there should be some limit.

Reply:
OK, I've implemented now a bounded stack based on a linked list.